### PR TITLE
Player class code cleanup + some performance fix

### PR DESCRIPTION
**Pull request by nerzhul**
### Commit 1:

* ClientEnvironment now uses UNORDERED MAP for active objects (performances)
* Use RemotePlayer and LocalPlayer everywhere it's possible
* Minor code style fixes
* Drop Client::getBreath() unused function

### Commit 2:
* Optimize ClientIface::getPlayerNames(): return const ref instead a copy of all names

### Commit 3:

RemotePlayer/LocalPlayer Player base class proper separation (code cleanup) (patch 3 of 4)

* remove IGameDef from Player class, only LocalPlayer has it now
* move many attributes/functions only used by LocalPlayer from Player to LocalPlayer
* move many attributes/functions only used by RemotePlayer from Player to RemotePlayer
* make some functions const
* hudGetHotbarSelectedImage now returns const ref (perf)
* RemotePlayer getHotbarSelectedImage now returns const ref (perf)
* various code style fixes

### Commit 4:
RemotePlayer/LocalPlayer Player base class proper separation (code cleanup) (patch 4 of 4)
* Move RemotePlayer code to its own cpp/header

### CODE REVIEW

File: src/environment.cpp [hunk 1]
```diff
@@ -2714,8 +2705,7 @@ u16 ClientEnvironment::addActiveObject(ClientActiveObject *object)
 		}
 		object->setId(new_id);
 	}
-	if(isFreeClientActiveObjectId(object->getId(), m_active_objects) == false)
-	{
+	if(!isFreeClientActiveObjectId(object->getId(), m_active_objects)) {
```
Comment (rubenwardy) [inline src/environment.cpp]: missed one

**Comment by nerzhul** [comment]
@est31 @Zeno- @rubenwardy @sfan5 can you look at this changeset ? I will not squash commits as each commit is a different task and require the previous commit.
I suggest you to look at each commit instead the whole changeset to see what changes on each step

File: src/clientiface.h [hunk 2]
```diff
@@ -453,7 +453,7 @@ class ClientInterface {
 	std::vector<u16> getClientIDs(ClientState min_state=CS_Active);
 
 	/* get list of client player names */
-	std::vector<std::string> getPlayerNames();
+	const std::vector<std::string> &getPlayerNames() const { return m_clients_names; }
```
Comment (rubenwardy) [inline src/clientiface.h]: may be worth declaring this inline (Although any compiler worth its salt would do this anyway, so meh

File: src/player.cpp [hunk 3]
```diff
@@ -309,11 +233,75 @@ void RemotePlayer::save(std::string savedir)
...
+	}
+
+	m_dirty = true;
+	//args.getS32("version"); // Version field value not used
+	std::string name = args.get("name");
+	strlcpy(m_name, name.c_str(), PLAYERNAME_SIZE);
+	setPitch(args.getFloat("pitch"));
+	setYaw(args.getFloat("yaw"));
+	setPosition(args.getV3F("position"));
+	try{
```
Comment (rubenwardy) [inline src/player.cpp]: a few missing spaces in this function

**Comment by rubenwardy** [comment]
Does the client have a concept of `isDead()`, or does it rely on the server telling it to die? If it does have that concept, then isDead() should be in player as it's a helper function related to `hp`, a player member

**Comment by nerzhul** [comment]
@rubenwardy for the isDead it's a serverside check. Client only triggers death when you do SendDeathScreen.

**Comment by rubenwardy** [comment]
As expected. All looks good (apart from last commit which I haven't checked) - just a few small nitpicks

**Comment by nerzhul** [comment]
last commit is only code move, no code change appart headers inclusion

File: src/clientiface.h (hunk 2 above)
Comment (nerzhul) [inline src/clientiface.h]: yes i agree, but with this mode we ensure we have the correct behaviour for constness

**Comment by nerzhul** [comment]
Fixed the code style issues in 4th commit

**Comment by nerzhul** [comment]
Add my own approval

File: src/remoteplayer.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	}
+
+	m_chat_message_allowance += time_passed * (m_setting_chat_message_limit_per_10sec / 8.0f);
+	if (m_chat_message_allowance > m_setting_chat_message_limit_per_10sec) {
+		m_chat_message_allowance = m_setting_chat_message_limit_per_10sec;
+	}
+
+	if (m_chat_message_allowance < 1.0f) {
+		infostream << "Player " << m_name
+				   << " chat limited due to excessive message amount." << std::endl;
```
Comment (SmallJoker) [inline src/remoteplayer.cpp]: Indentation not correct

File: src/server.cpp [hunk 5]
```diff
@@ -3199,7 +3188,7 @@ void Server::deleteParticleSpawner(const std::string &playername, u32 id)
 
 	u16 peer_id = PEER_ID_INEXISTENT;
 	if (playername != "") {
-		Player* player = m_env->getPlayer(playername.c_str());
+		RemotePlayer* player = m_env->getPlayer(playername.c_str());
```
Comment (SmallJoker) [inline src/server.cpp]: Pointer on the wrong side (above too)

**Comment by nerzhul** [comment]
Code style fixed, thanks @SmallJoker

**Pull request merged**